eth_account/hdaccount/deterministic.py

```diff
@@ -0,0 +1,217 @@
+  form: (0x02 or 0x03) || ser_256(x), where the header byte depends on the parity of the
+  omitted y coordinate.
+* parse_256(p): interprets a 32-byte sequence as a 256-bit number, most significant byte first.
+
+"""
+# Additional notes:
+# - This algorithm only implements private parent key → private child key CKD function,
+#   as it is unnecessary to the HD key derivation functions used in this library to implement
+#   the other functions (as Ethereum uses an Account-based system)
+# - Unlike other libraries, this library does not use Bitcoin key serialization, because it is
+#   not intended to be ultimately used for Bitcoin key derivations. This presents a simplified API.
+import hashlib
+import hmac
+from typing import (
+    Tuple,
+    Union,
+)
+
+from eth_keys import (
+    keys,
+)
+from eth_utils import (
+    to_int,
+)
+from hexbytes import (
+    HexBytes,
+)
+
+SECP256K1_N = int("FFFFFFFF_FFFFFFFF_FFFFFFFF_FFFFFFFE_BAAEDCE6_AF48A03B_BFD25E8C_D0364141", 16)
+
+
+def hmac_sha512(chain_code: bytes, data: bytes) -> bytes:
+    """
+    As specified by RFC4231 - https://tools.ietf.org/html/rfc4231
+    """
+    return hmac.new(chain_code, data, hashlib.sha512).digest()
+
+
+def ec_point(pkey: bytes) -> bytes:
+    """
+    Compute `point(p)`, where `point` is ecdsa point multiplication
+
+    Note: Result is ecdsa public key serialized to compressed form
+    """
+    return keys.PrivateKey(HexBytes(pkey)).public_key.to_compressed_bytes()
+
+
+class SoftNode(int):
+    """
+    Soft node (unhardened), where value = index
+    """
+    def __new__(cls, index):
+        obj = int.__new__(cls, index)
+        obj.index = index
+        return obj
+
+    def __add__(self, other: int):
+        return self.__class__(self.index + other)
+
+    def serialize(self) -> bytes:
+        assert 0 <= self < 2**32
+        return self.to_bytes(4, byteorder="big")
+
+    def encode(self) -> str:
+        return str(self.index)
+
+    def __repr__(self):
+        return f"SoftNode({self.index})"
+
+
+class HardNode(int):
+    """
+    Hard node, where value = index + BIP32_HARDENED_CONSTANT
+    """
+    def __new__(cls, index):
+        obj = int.__new__(cls, index + 0x80000000)  # 2**31, BIP32 "Hardening constant"
+        obj.index = index
+        return obj
+
+    def __add__(self, other: int):
+        return self.__class__(self.index + other)
+
+    def serialize(self) -> bytes:
+        assert 0 <= self < 2**32
+        return self.to_bytes(4, byteorder="big")
+
+    def encode(self):
+        return str(self.index) + "'"
+
+    def __repr__(self):
+        return f"HardNode({self.index})"
+
+
+def decode_node(node: str) -> Union[SoftNode, HardNode]:
+    if len(node) < 1:
+        raise ValueError("Cannot use empty string")
+    if node[-1] in ("'", "H"):
+        return HardNode(int(node[:-1]))
+    else:
+        return SoftNode(int(node))
+
+
+def derive_child_key(
```

Comment:
I think it makes sense to group all of these functions that are just floating around in here (`derive_child_key`, `decode_node`, `ec_point`, and  maybe `hmac_sha512`) inside of a class. Maybe call it `HDWallet` or something (but I can't really think of a good name). What do you think?

Reply:
I was thinking of making `decode_node` a staticmethod of a `Node` base class (per your suggestion about), and making `ec_point` and `hmac_sha512` into functions in a `_utils.py` file under this module.

`derive_child_key` may eventually be accessible to the outside world though, as it may be useful for #25 & #37

Reply:
👍 Yeah, like the Node base class and putting ec_point and hmac_sha512 into a _utils file.